python/ray/data/_internal/arrow_ops/transform_pyarrow.py

```diff
@@ -57,6 +57,66 @@ def sort(table: "pyarrow.Table", sort_key: "SortKey") -> "pyarrow.Table":
     return take_table(table, indices)
 
 
+def _create_empty_table(schema: "pyarrow.Schema"):
+    import pyarrow as pa
+
+    arrays = [pa.array([], type=t) for t in schema.types]
+
+    return pa.table(arrays, schema=schema)
+
+
+def hash_partition(
+    table: "pyarrow.Table",
+    *,
+    hash_cols: List[str],
+    num_partitions: int,
+) -> Dict[int, "pyarrow.Table"]:
+    """Hash-partitions provided Pyarrow `Table` into `num_partitions` based on
+    hash of the composed tuple of values from the provided columns list
+
+    NOTE: Since some partitions could be empty (due to skew in the table) this returns a
+          dictionary, rather than a list
+    """
+
+    import numpy as np
+
+    assert num_partitions > 0
+
+    if table.num_rows == 0:
+        return {}
+    elif num_partitions == 1:
+        return {0: table}
+
+    projected_table = table.select(hash_cols)
+
+    partitions = np.zeros((projected_table.num_rows,))
+    for i in range(projected_table.num_rows):
+        _tuple = tuple(c[i] for c in projected_table.columns)
+        partitions[i] = hash(_tuple) % num_partitions
```

Comment:
Unfortunately `hash` is not stable across python VMs, therefore this will not work as expected.
https://docs.python.org/3/reference/datamodel.html#object.__hash__

Not sure if it's feasible to require the setting of [PYTHONHASHSEED](https://docs.python.org/3/using/cmdline.html#envvar-PYTHONHASHSEED) for this to work.

I left one suggestion to address this here, using hashlib: https://github.com/ray-project/ray/issues/49133#issuecomment-2827824693

FYI I also tryed to use [pandas.hash_array](https://github.com/pandas-dev/pandas/blob/v2.2.3/pandas/core/util/hashing.py#L233) but it's not filling the partition space correctly (e.g. for target of 20 partitions, only 4 would have values).

Reply:
In this case actually, we only run hash on Arrow blocks columns, which actually override their hash hence not requiring the deserialization to Python.

> Edit: I also think we should allow for exact partition assignment via a custom partitioner, just like spark, which is basically a function that return the target partition index

We will definitely add that in the future. I'd recommend creating a ticket for this (feel free to paste it here as well)